checker/cost.go

```diff
@@ -0,0 +1,536 @@
+package checker
+
+import (
+	"fmt"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"math"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/containers"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// CostEstimator estimates the sizes of variable length input data and the costs of functions.
+type CostEstimator interface {
+	// EstimateSize returns the worst case size possible for the given Expr, or nil if
+	// the estimator has no estimate to provide. EstimateSize is only called for Exprs where
+	// CEL does not know the size; EstimateSize is not called for literals defined inline in CEL.
+	EstimateSize(element Expr) *int64
```

Comment:
Consider returning a `*SizeEstimate` here. though the size feels more like estimating the bounds. Perhaps, `EstimateBounds`?

Another option is to just reuse `CostEstimate` as the result from both calls, though the name might need to change if this were the case.